02-peer-protocol.md

```diff
@@ -49,11 +68,471 @@ transaction is confirmed are also not persistent - until you know the script
+
+Upon successful exchange of `tx_complete` messages, both nodes
+should construct the transaction and proceed to the next portion of the
+protocol. For channel establishment v2, exchanging commitment transactions.
+
+For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).
+
+The maximum inputs and outputs are capped at 252. This effectively fixes
+the byte size of the input and output counts on the transaction to one (1).
+
+### The `tx_signatures` Message
+
+1. type: 71 (`tx_signatures`)
+2. data:
+    * [`channel_id`:`channel_id`]
+    * [`sha256`:`txid`]
+    * [`u16`:`num_witnesses`]
+    * [`num_witnesses*witness_stack`:`witnesses`]
+
+1. subtype: `witness_stack`
+2. data:
+    * [`u16`:`num_witness_elements`]
+    * [`num_witness_elements*witness_element`:`witness_elements`]
+
+1. subtype: `witness_element`
+2. data:
+    * [`u16`:`len`]
+    * [`len*byte`:`witness_data`]
+
+#### Requirements
+
+The sending node:
+  - if it has the lowest total satoshis contributed, as defined by total
+    `tx_add_input` values, or both peers have contributed equal amounts
+    but it has the lowest `node_id` (sorted lexicographically):
+    - MUST transmit their `tx_signatures` first
+  - MUST order the `witnesses` by the `serial_id` of the input they
+    correspond to
+  - `num_witnesses`s MUST equal the number of inputs they added
+
+The receiving node:
+  - MUST fail the negotiation if:
+    - the message contains an empty `witness_stack`
+    - the number of `witnesses` does not equal the number of inputs
+      added by the sending node
+    - the `txid` does not match the txid of the transaction
+    - the witnesses are non-standard
+  - SHOULD apply the `witnesses` to the transaction and broadcast it
+  - MUST reply with their `tx_signatures` if not already transmitted
+
+#### Rationale
+
+A strict ordering is used to decide which peer sends `tx_signatures` first.
+This prevents deadlocks where each peer is waiting for the other peer to
+send `tx_signatures`, and enables multiparty tx collaboration.
+
+`witness_data` is the data for a witness element in a witness stack, not
+prefixed with its length (since it is already specified in the `len` field).
+
+While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
+it is possible for the fee for the exchanged witness data to be underpaid.
+It is the responsibility of the sending peer to correctly account for the
+required fee, e.g. a multisig witness stack whose weight exceeds 107.
+
+### The `tx_init_rbf` Message
+
+This message initiates a replacement of the transaction after it's been
+completed.
+
+1. type: 72 (`tx_init_rbf`)
+2. data:
+   * [`channel_id`:`channel_id`]
+   * [`u32`:`locktime`]
+   * [`u32`:`feerate`]
+   * [`tx_init_rbf_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `tx_init_rbf_tlvs`
+2. types:
+    1. type: 0 (`funding_output_contribution`)
```

Comment:
If `funding_output_contribution` is missing, do we assume the contribution is the same as before?

Reply:
If `funding_output_contribution` is missing, it means the peer is not contributing to the channel (it's equivalent to having set a value of `0`). That should be explicitly mentioned.

Reply:
Fixed in https://github.com/niftynei/lightning-rfc/pull/10

Reply:
merged!